tools/discovery/discovery.go

```diff
@@ -0,0 +1,131 @@
+package discovery
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"github.com/xiekeyang/oci-discovery/tools/object"
+)
+
+var (
+	defaultTrans = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+)
+
+func DiscoveryHandler(context *cli.Context) error {
+	var (
+		name  = context.Args()[0]
+		roots []v1.Descriptor
+	)
+
+	v, err := paramsParser(name)
+	if err != nil {
+		return err
+	}
+
+	engines, err := refEnginesFetching(v)
```

Comment:
This needs to [iterate over protocols][1] and [walk ancestor hosts][2] in case the first attempted host/protocol does not resolve the name.

[1]: https://github.com/xiekeyang/oci-discovery/blame/44ec3cf3113e29a743ad04220ccb7ff5197dab2a/oci_discovery/ref_engine_discovery/__init__.py#L34
[2]: https://github.com/xiekeyang/oci-discovery/blame/44ec3cf3113e29a743ad04220ccb7ff5197dab2a/oci_discovery/ref_engine_discovery/__init__.py#L35

Reply:
Yes, I skip `protocol` entry, because I not figure out the nice `structure` or `interface` building. I planed to do it next step.